### ISSUE TITLE
Is not support Windows?

### ISSUE THREAD

**Issue by ZHANITEST** [i2p/libsam3#7]
Hi! 😄

I just wanna use libsam3.c on D programming language with the C FFI.
But libsam3.c used some POSIX headers(ex. sys/netdb).
Is not any plan for support windows?

Regards,

**Comment by sehraf** [comment]
Windows build with mingw64 fails:
```
$ LANG=C make
cc -Wall -g -O2 -std=gnu99 -c src/libsam3/libsam3.c -o src/libsam3/libsam3.o
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:56:8: error: unknown type name 'ssize_t'
   56 | extern ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize,
      |        ^~~~~~~
src/libsam3/libsam3.h:59:8: error: unknown type name 'ssize_t'
   59 | extern ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize);
      |        ^~~~~~~
src/libsam3/libsam3.h:283:8: error: unknown type name 'ssize_t'
  283 | extern ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize);
      |        ^~~~~~~
src/libsam3/libsam3.h:299:8: error: unknown type name 'ssize_t'
  299 | extern ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |        ^~~~~~~
src/libsam3/libsam3.c:187:9: error: conflicting types for 'sam3tcpReceiveEx'
  187 | ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize, int allowPartial) {
      |         ^~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:56:16: note: previous declaration of 'sam3tcpReceiveEx' was here
   56 | extern ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize,
      |                ^~~~~~~~~~~~~~~~
src/libsam3/libsam3.c:213:9: error: conflicting types for 'sam3tcpReceive'
  213 | ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize) {
      |         ^~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:59:16: note: previous declaration of 'sam3tcpReceive' was here
   59 | extern ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize);
      |                ^~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:12:
src/libsam3/libsam3.c: In function 'xstrtokend':
src/libsam3/libsam3.c:415:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  415 |   while (*s && isspace(*s))
      |                        ^~
src/libsam3/libsam3.c:429:26: warning: array subscript has type 'char' [-Wchar-subscripts]
  429 |       } else if (isspace(*s)) {
      |                          ^~
src/libsam3/libsam3.c: In function 'sam3ParseReply':
src/libsam3/libsam3.c:446:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  446 |   while (*p && isspace(*p))
      |                        ^~
src/libsam3/libsam3.c:459:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  459 |   while (*e && isspace(*e))
      |                        ^~
src/libsam3/libsam3.c:466:26: warning: array subscript has type 'char' [-Wchar-subscripts]
  466 |     while (*p && isspace(*p))
      |                          ^~
src/libsam3/libsam3.c: At top level:
src/libsam3/libsam3.c:1139:9: error: conflicting types for 'sam3DatagramReceive'
 1139 | ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize) {
      |         ^~~~~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:283:16: note: previous declaration of 'sam3DatagramReceive' was here
  283 | extern ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize);
      |                ^~~~~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:12:
src/libsam3/libsam3.c: In function 'sam3DatagramReceive':
src/libsam3/libsam3.c:1171:27: warning: array subscript has type 'char' [-Wchar-subscripts]
 1171 |     if (!v[0] || !isdigit(*v)) {
      |                           ^~
src/libsam3/libsam3.c:1177:26: warning: array subscript has type 'char' [-Wchar-subscripts]
 1177 |     while (*v && isdigit(*v)) {
      |                          ^~
src/libsam3/libsam3.c: At top level:
src/libsam3/libsam3.c:1206:9: error: conflicting types for 'sam3Base32Encode'
 1206 | ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |         ^~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:299:16: note: previous declaration of 'sam3Base32Encode' was here
  299 | extern ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |                ^~~~~~~~~~~~~~~~
make: *** [Makefile:37: src/libsam3/libsam3.o] Error 1
```

**Comment by eyedeekay** [comment]
@sehraf Are you in a position to give: https://github.com/i2p/libsam3/pull/9 a try?

**Comment by sehraf** [comment]
yes, but same result.
I'm using MSYS2 as described [here ](http://docs.retroshare.net/en/latest/developer/compilation/#compilation-on-windows)

```
$ LANG=C make
cc -Wall -g -O2 -std=gnu99 -c src/libsam3/libsam3.c -o src/libsam3/libsam3.o
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:56:8: error: unknown type name 'ssize_t'
   56 | extern ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize,
      |        ^~~~~~~
src/libsam3/libsam3.h:59:8: error: unknown type name 'ssize_t'
   59 | extern ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize);
      |        ^~~~~~~
src/libsam3/libsam3.h:283:8: error: unknown type name 'ssize_t'
  283 | extern ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize);
      |        ^~~~~~~
src/libsam3/libsam3.h:299:8: error: unknown type name 'ssize_t'
  299 | extern ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |        ^~~~~~~
src/libsam3/libsam3.c:202:9: error: conflicting types for 'sam3tcpReceiveEx'
  202 | ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize, int allowPartial) {
      |         ^~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:56:16: note: previous declaration of 'sam3tcpReceiveEx' was here
   56 | extern ssize_t sam3tcpReceiveEx(int fd, void *buf, size_t bufSize,
      |                ^~~~~~~~~~~~~~~~
src/libsam3/libsam3.c:228:9: error: conflicting types for 'sam3tcpReceive'
  228 | ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize) {
      |         ^~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:59:16: note: previous declaration of 'sam3tcpReceive' was here
   59 | extern ssize_t sam3tcpReceive(int fd, void *buf, size_t bufSize);
      |                ^~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:12:
src/libsam3/libsam3.c: In function 'xstrtokend':
src/libsam3/libsam3.c:430:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  430 |   while (*s && isspace(*s))
      |                        ^~
src/libsam3/libsam3.c:444:26: warning: array subscript has type 'char' [-Wchar-subscripts]
  444 |       } else if (isspace(*s)) {
      |                          ^~
src/libsam3/libsam3.c: In function 'sam3ParseReply':
src/libsam3/libsam3.c:461:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  461 |   while (*p && isspace(*p))
      |                        ^~
src/libsam3/libsam3.c:474:24: warning: array subscript has type 'char' [-Wchar-subscripts]
  474 |   while (*e && isspace(*e))
      |                        ^~
src/libsam3/libsam3.c:481:26: warning: array subscript has type 'char' [-Wchar-subscripts]
  481 |     while (*p && isspace(*p))
      |                          ^~
src/libsam3/libsam3.c: At top level:
src/libsam3/libsam3.c:1154:9: error: conflicting types for 'sam3DatagramReceive'
 1154 | ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize) {
      |         ^~~~~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:283:16: note: previous declaration of 'sam3DatagramReceive' was here
  283 | extern ssize_t sam3DatagramReceive(Sam3Session *ses, void *buf, size_t bufsize);
      |                ^~~~~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:12:
src/libsam3/libsam3.c: In function 'sam3DatagramReceive':
src/libsam3/libsam3.c:1186:27: warning: array subscript has type 'char' [-Wchar-subscripts]
 1186 |     if (!v[0] || !isdigit(*v)) {
      |                           ^~
src/libsam3/libsam3.c:1192:26: warning: array subscript has type 'char' [-Wchar-subscripts]
 1192 |     while (*v && isdigit(*v)) {
      |                          ^~
src/libsam3/libsam3.c: At top level:
src/libsam3/libsam3.c:1221:9: error: conflicting types for 'sam3Base32Encode'
 1221 | ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |         ^~~~~~~~~~~~~~~~
In file included from src/libsam3/libsam3.c:10:
src/libsam3/libsam3.h:299:16: note: previous declaration of 'sam3Base32Encode' was here
  299 | extern ssize_t sam3Base32Encode(char *dest, size_t destsz, const void *srcbuf,
      |                ^~~~~~~~~~~~~~~~
make: *** [Makefile:37: src/libsam3/libsam3.o] Error 1
```

**Comment by eyedeekay** [comment]
Crap, I am not, cross-compiling from Debian. It's working for me from this direction, but I think I must be wrong, this is all POSIX stuff that Windows apparently doesn't have any clue about. Give me a little while to get MSYS2 set up on my machine so I can see what you see.

**Comment by eyedeekay** [comment]
Added a typedef for ssize_t on Windows, that will resolve some of your errors but possibly not all

**Comment by sehraf** [comment]
Still getting the same errors regarding `ssize_t`. This is basically the same trouble i had with the retrosahre pr. I'm not even sure if this is a problem of the build setup / environment or not.

**Comment by eyedeekay** [comment]
Darn, I thought that would get us closer. Once QT gets installed on this laptop I'll be able to see a little more. How much do you know about the C-Preprocessor? I may have gotten the #ifdef on libsam3.h line 17 wrong for your environment, we should be setting that(I think) so that it will affect every Windows build.

**Comment by sehraf** [comment]
No much, i'm mostly familiar with simple macros. 😞

**Comment by eyedeekay** [comment]
Same here. No worries, I'm sure it'll actually seem very obvious once I can see what's happening, I have a hunch already. I'm at about 10% QT installation now, but I pushed a change to the windows-libsam3 branch which does the #ifdef differently and should enable the typedef on mingw builds on Windows if you want to try that while I wait.

**Comment by eyedeekay** [comment]
So after 5 hours of getting QT installed... it builds for me. No ssize_t typedef errors at all.

**Comment by sehraf** [comment]
Thanks for clarifying this! So it is my setup...

**Comment by eyedeekay** [comment]
No problem, let me know if there's anything you discover that I can add to the docs to help other people configure their projects. I pretty straightforwardly followed the instructions from the RetroShare readthedocs except for replacing the extant libsam3 library with the windows-libsam3 branch.

**Comment by sehraf** [comment]
Got it compiled! Had to remove libsam3a.c from the `SRCS` though

```makefile
SRCS := \
	src/libsam3/libsam3.c \
	#src/libsam3a/libsam3a.c
```

RetroShare fails linking now, thouhg this seems to be something else #onlyWindowsThings

**Comment by eyedeekay** [comment]
You should only need to set the winsocks linker flags, I added them to the README.md in windows-libsam3. I haven't finished porting libsam3a yet, but I will only be another day or two on it I think.

**Comment by eyedeekay** [comment]
@sehraf You should be able to build without needing to remove libsam3a now. I'm not sure the approach I've taken is valid for the SOCK_NONBLOCKING and SOCK_CLOEXEC flags though so I need to check that tomorrow. I expect I'll have Windows support merged by the evening here.

**Comment by eyedeekay** [comment]
Closed by #9

**Comment by sehraf** [comment]
Got it finally working! Thanks for adding windows support!
